# Исследование надежности заемщиков


### Описание проекта
Заказчик — кредитный отдел банка. Нужно разобраться, влияет ли семейное положение и количество детей клиента на факт погашения кредита в срок. Входные данные от банка — статистика о платёжеспособности клиентов.

Результаты исследования будут учтены при построении модели кредитного скоринга — специальной системы, которая оценивает способность потенциального заёмщика вернуть кредит банку.
### Описание данных
- `children` — количество детей в семье
- `days_employed` — общий трудовой стаж в днях
- `dob_years` — возраст клиента в годах
- `education` — уровень образования клиента
- `education_id` — идентификатор уровня образования
- `family_status` — семейное положение
- `family_status_id` — идентификатор семейного положения
- `gender` — пол клиента
- `income_type` — тип занятости
- `debt` — имел ли задолженность по возврату кредитов
- `total_income` — ежемесячный доход
- `purpose` — цель получения кредита

## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [14]:
data.duplicated().sum()

54

In [15]:
data = data.drop_duplicates()

**Задание 15. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**

In [16]:
data['education'] = data['education'].str.lower()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

## Исследуйте данные и ответьте на вопросы

**Задание 19. Есть ли зависимость между количеством детей и возвратом кредита в срок?**

In [22]:
# напишем функцию, которая будет подсчитывать долю(%) людей не вернувших кредит в срок - debtors_share_% - в зависимости от интересующего нас параметра (названия колонки датафрейма)
def parameter_debt_relation(parameter):
    parameter_debt_relation = data.groupby(parameter).agg({'debt': ['sum', 'count']})
    parameter_debt_relation['sample_size'] = parameter_debt_relation['debt']['count']
    parameter_debt_relation['debtors_share_%'] = (parameter_debt_relation['debt']['sum']/parameter_debt_relation['debt']['count']*100)\
                                                    .round(2)
    return(parameter_debt_relation.sort_values('debtors_share_%', ascending=False).drop(columns='debt'))

In [23]:
parameter_debt_relation('children')

sample_size debtors_share_%
                                    
children                            
4                 41            9.76
2               2052            9.45
1               4809            9.23
3                330            8.18
0              14107            7.54
5                  9            0.00

**Вывод:** Исходя из представленных данных, наибольшая доля не вернувших кредит в срок приходится на группу людей, имеющих от 1-го до 4-х детей (для всех обозначенных категорий в среднем 8-10% не возвращают кредит вовремя). Наименьший риск, связанный с неуплатой кридита, у заемщиков либо не имеющих детей вообще, либо имеющих 5 детей, доля невозврата кредита 7.54% и 0% соответственно. 
***
*Однако, стоит учитывать, что в нашем случае выборка не сбалансирована и скошена в сторону заемщиков без детей, поэтому какие-то выводы относительно заемщиков с большим количеством детей делать сложно.*

**Задание 20. Есть ли зависимость между семейным положением и возвратом кредита в срок?**

In [24]:
parameter_debt_relation('family_status')

sample_size debtors_share_%
                                                 
family_status                                    
Не женат / не замужем        2796            9.76
гражданский брак             4146            9.29
женат / замужем             12266            7.56
в разводе                    1189            7.06
вдовец / вдова                951            6.62

**Вывод:** Доля не выплативших кредит в срок заемшиков из категории не состоящих в браке (группа "не женат / не замужем") и заемщиков состоящих в гражданском браке (группа "гражданский брак") наибольшая (9.76% и 9.29% соответственно). Люди из категорий "женат / замужем",	"в разводе", "вдовец / вдова" являются наиболее надежными с точки зрения их кредитной способности (доля должников среди этих групп ниже 8%).

**Задание 21. Есть ли зависимость между уровнем дохода и возвратом кредита в срок?**

In [25]:
parameter_debt_relation('total_income_category')

sample_size debtors_share_%
                                                 
total_income_category                            
E                              22            9.09
C                           15938            8.49
A                              25            8.00
B                            5014            7.06
D                             349            6.02

**Вывод:** Кредитополучатели из категорий "E" и "C" (доход до 30000 и 50001-200000 соответственно) являются наименее надежными с точки зрения рисков неуплаты кредита. Группы кредитополучателей с доходом от 200001 и более (категории "A" и "B") являются надежными, доли не вернувших кредит в срок, согласно представленным данным, 8% и 7.06% соответственно. Наименьший риск неуплаты кридита в срок определен для заемщиков категории "D" (доход 30001–50000), доля должников в этой группе составляет 6.02%.
***
*Учтем, что в данных плохо представлены группы заемщиков из категорий "E" и "A", что  затрудняет интерперетацию данных и формульровку выводов.*

**Задание 22. Как разные цели кредита влияют на его возврат в срок?**

In [26]:
parameter_debt_relation('purpose_category')

sample_size debtors_share_%
                                                    
purpose_category                                    
операции с автомобилем          4281            9.34
получение образования           3989            9.25
проведение свадьбы              2324            7.87
операции с недвижимостью       10754            7.25

**Вывод:** Среди всех групп, сформированных на основании цели получения кредита, доля должников наибольшая для категорий "операции с автомобилем" и "получение образования" (на основании представленных данных - 9.34% и 9.25% соответственно). В свою очередь, кредиты на "проведение свадьбы" или для "операций с недвижимостью" обладают наименьшим риском (доли должников 7.87% и 7.25%).

In [27]:
data\
    .pivot_table(index='purpose_category', values='debt', aggfunc='mean')\
    .mul(100)\
    .round(2)\
    .rename(columns={'debt': 'debtors_share_%'})\
    .sort_values(by='debtors_share_%', ascending=False)

debtors_share_%
purpose_category                         
операции с автомобилем               9.34
получение образования                9.25
проведение свадьбы                   7.87
операции с недвижимостью             7.25

**Задание 23. Приведите возможные причины появления пропусков в исходных данных.**

*Ответ:* 
- Информация о трудовом стаже и доходе для определенных заемщиков скрыта
- Неправильный ввод данных / выгрузка из базы 

**Задание 24. Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.**

*Ответ:* 
- Для количественных переменны пропуски обычно заполняются *характерными значениями*. Предполагается, что данные значения максимально приближены к реальным. 
- Медиана находится в середине упорядоченного набора значений, то есть это такое число, что половина из элементов набора не меньше него, а другая половина не больше. Медиана более устойчива к появлению сильно завышенных или сильно заниженных значений в пределах группы данных (выбросов).

## Общий вывод.

*Решение, стоит ли выдавать кредит или нет (а также какой лимит кредитования и проценты установить), банк принимает в зависимости от классификации клиента по группам риска. Таким образом существует необходимость выделять те характеристики, которые наиболее тесно связаны с ненадежностью или, наоборот, с надежностью клиента.*
***
В результате представленного исследования и анализа ряда параметров, потенциально определяющих кредитную способность клиента, были выделены ключевые категории, которые являются "ненадежными" с точки зрения возможных рисков просрочки платежа по кредиту. К данным категориям относятся:
1. По количеству детей:
  - лица, имеющие от 1 до 4 детей
2. По семейному положению:
  - неженатые/незамужние
  - лица, состоящие в гражданском браке
3. По уровню дахода:
  - лица с доходом до 30000
  - лица с доходом 50001–200000
4. По цели кредита:
  - операции с автомобилем
  - получение образования

Попадение потенциального кредитополучателя в одну или несколько "ненадежных" категорий может расцениваться как повод для отказа в предоставлении кредита. Полученные данные могут быть использованы для постороения системы кредитного скоринга.